### DISCUSSION TITLE
Sampling values to limit output bandwidth

### DISCUSSION THREAD

**Discussion by damiensellier** [RomanKubiak/ctrlr#230] [Q&A]
Hi everybody,

A lot of old devices cannot handle more than 1 incoming message every 50-100ms, or they overload, stop processing and bug.
I was wondering if it is possible to only consider 1 value for a parameter change every t  _ms_ to limit the amount of datas sent to the controlled device ?
Let's say, I change the modulator value from 0 to 127 in 1 sec, in a linear way, it's going to send a message for every 128 steps in 1 second (almost 1 sysex message every 8ms) which is useless and overloading. Is there a way to only take in account the current value every 50ms for ex, like in a sampling operation?

Thanks for your help

**Comment by Godlike-Productions** [comment]
I've done this on a couple of panels.

The easiest way is just to set the MIDI rate at the global level of the panel (panelMidiGlobalDelay)

A more refined procedure is this


Pick your minimum sampling time in ms.

Stop your modulator from sending MIDI.

Set a timerCallback at your sampling time. In the procedure, sample the modulator's value and send it using sendMidiMessageNow

Make sure you have a way to stop the loop (I use a control mode button)

**Reply by DRBAUS** [reply]
I am working on a Panel for the Ensoniq EPS-16 Plus/ASR-10 and these samplers are terrible in handling sysex. 
Each Modulator value sends Instrument, Layer and Parameter values, it has to,  and the machines are using an internal delay which means that scrolling from value 0 to 99 takes a long time to arrive at value 99.

My best way of dealing with this is to make the Panel slider-less, so no visible/physical sliders, only value input boxes. This calls for a "value input method" of editing. Small adaptations to values, like 3 or 4 values up or down, can still be done with the mouse wheel.

This works fast as well but yeah, there may be better ways to deal with this. I am not sure yet that I understand how to implement Godlike-Productions' example. I will keep following this thread and look at it when I continue working on the Panel.

**Reply by dobo365** [reply]
_Out of topic_
FYI, @DRBAUS, I have still not succeeded to send any sysex dump to my TS10 (sounddiver or MidiOX). I can receive dumps fine.
Now that I have a Gotek with all files on it, it is less important but anyway, I hope to solve that one day ;-)

**Reply by damiensellier** [reply]
I am working on finishing my Ensoniq DP4 panel and go through all the remaining optimisation and bugs.
This sampling value is one of them. I recorded an automation of a value for a minute in cubase and with panelMidiGlobalDelay set at 50ms between messages, ctrlr was still sending values 2 minutes after the rec stopped :)
SO it's definitely not a viable solution. I have to find a way to include the variable T in the equation for all the values considered at the modulator level or at the sendmidimessage one.
I really need to explore this "timerCallback" function because I never came across it. 
I'll check the JUCE library tonight, I'm pretty sure there's something about that, but as always, there's no guarantee Roman implemented it in Ctrlr

if it's possible to set a global sampleRate, this would be easy with JUCE setNextValue() and we could use SmoothedValue() for clean smooth curves.

A global parameter in ctrlr itself would be great like a panelMidiGlobalResolution

**Reply by DRBAUS** [reply]
Just to understand better: what do you mean by Sampling Value? What does this do in the DP4? Why do you need to automate this?

**Reply by damiensellier** [reply]
@DRBAUS 
What you propose is ok but it's like reducing the parameter value scale from 128 to a smaller value scale then multiplying it back to get the equivalent to a 128 value scale. You reduce the flow but you cut off the precision/depth of the value. I think it would be better to keep the precision/depth of the value but limit the amount of data per second. So it's time related not value based.    It's a bit like reducing audio Bit depth vs reducing audio sampling frequency to get a smaller data flow. The result is not the same.

So here I'm not talking  sampling value related to Audio  sampling frequency(44.1KHz, 48KHz etc) of the DP4 but I am talking about a timing resolution, like midi pulses per quarter note (PPQN) for exemple. 
Let's say in 1 sec you move your pitch wheel from 0 to 16384 via sysex message imagine the flow of linear messages sent.... When I talk about sampling I talk about capturing and holding a value at instant t where the next value to be considered will be at t+P  with P the period between 2 samples.   so with a "sampleRate" of 10Hz it will consider only 10 values in 1 second, 1 value every 100ms which is very acceptable by the old gears. I tried sending messages with a globalDelay of 100ms it's acceptable, but under 80ms, I start getting sysex message errors. 
I'm looking for a way in Ctrlr to introduce time in the equation

**Reply by DRBAUS** [reply]
Ah...yeah...I understand.

I don't know if this will help you in any way but.... In the Panel for the Ensoniq EPS-16 Plus I made a few labels to set the sample start, sample end, loop start and loop end points. Since these values can be very large and I had no idea how to deal with that the guys helped me with a method.

You can find it here:

https://github.com/RomanKubiak/ctrlr/discussions/129

Maybe this can be of use to you as well.

**Reply by damiensellier** [reply]
@DRBAUS thanks but I don't need to translate values it's really about introducing the "timing" in the formulas.

**Reply by Godlike-Productions** [reply]
For the EPS and ASR -10 can you send Data (CC6) to quickly change the value of the selected parameter (ie send first value as the sysex string to select it, and then update the value with CC6).  That's pretty much how it works from the front panel if I recall correctly.

The method I use is equivalent to thinning the data.  It doesn't send every single value, it just samples the slider state as predetermined intervals and sends those.  You lose no resolution of the control, and if you get the timing right, you don't lose MIDI resolution as you are sending it at close to the maximum rate the synth can handle.  You just send your MIDI messages at a rate that the synth can handle, skipping data in between.

Also, I do a check to see if the data has changed since the last thing sent.  If it hasn't then I don't send anything, so it's a really efficient way or reducing MIDI traffic.

I'll see if I can pull a code example together. My implementation is slightly different because I'm sampling every control on the panel, assembling a sysex dump and sending that to the synth's edit buffer every 200ms, so mine is global to the panel.

I'm trying to think how it would work with multiple parameters that are automated, each on their own timer.  You'd need a way of scheduling them, or prioritizing them.

The other option is to look at what I just did with the AlphaWheel (different thread).  I only send the last value of a moved slider, but at the moment it only works where the controller is springed (ie returns to a value).  Might be able to adapt with a timer somehow to send the last value, but need to work out a way to detect that its stopped moving. Maybe keep refreshing a timer somehow. and send when the timer expires.

**Reply by damiensellier** [reply]
@Godlike-Productions If I understand your message, it's close to my fallback method. Worst cas scenario, I can map values to CC controllers (x2) and you can feed the machine with midi CC values without any problem. But I would prefer just sending sysex parameter value changes via sysex since it's more straight forward within my interface and remove the parameter to CC controler assignement step. 
![Capture d’écran, le 2021-03-28 à 18 26 20](https://user-images.githubusercontent.com/55164385/112770143-44db9c00-8ff3-11eb-920b-6857bc71aceb.png)

**Reply by Godlike-Productions** [reply]
I haven't read the DP4+ Midi spec for a while. I can't remember if it support NRPN's, but the spec is pretty comprehensive.

NRPN's are much more efficient in terms of MIDI bandwidth than SYSEX generally (except if the sysex is less than 8 bytes).  NRPN's also support running messages, which is handy.

**Reply by damiensellier** [reply]
@Godlike-Productions When you say " it just samples the slider state as predetermined intervals and sends those." In Ctrlr how do you introduce this interval of time in the formulas?

**Reply by Godlike-Productions** [reply]
You need to do it in LUA.

I have a button that starts control mode (I disable other MIDI dump functionality while in control mode)

It starts a timerCallback

In the timerCallback function it samples and sends the data.  Inside that function is a check for the state of the control mode button that can stop the timer (exiting the timerCallback).

The benefit of this is that it doesn't slow down the panel.  You can move the controls on the panel while the callback is waiting to send it's "interrupt" back.

**Reply by damiensellier** [reply]
Sure I understand, now the time is taken in consideration. I'll check the timerCallback tonight and see how I can adjust it to my scripts. thanks !

**Reply by Godlike-Productions** [reply]
Here's the code I use.

This launches when I press the controlModeButton

```
controlModeMethod = function(--[[ CtrlrModulator --]] mod, --[[ number --]] value, --[[ number --]] source)

	-- No action if the panel is in bootstrap or program states
	if not isPanelReady() then -- this is a seperate function to check panel state.
		return
	end

-- Start  Timer Loop
timer:setCallback (1,timerCallback)
timer:startTimer (1,midiDelayTime) -- need to set this to variable. I use a label to get this

end
```

This is a little bit of my timerCallback

```

function timerCallback(timerId)

-- this function formats midiBuffer as a Edit Buffer Load. Patch data starts at byte 7. Don't use for 
-- patch dumps to specific patch locations

-- byte's 07 and 08 are the patch protected byte. 0xff is protected, 0x00 is not-protected. Use 0x00

	-- Load program timer
	if timerId == 1 then
		if lastMidiBuffer == {0x05} then -- this can be removed. It's handled differently now.
		--	timer:stopTimer(timerId)
			return
		else
		-- Read VCO2 Range Slider
			vco2RSliderValue = panel:getModulatorByName("vco2RangeSlider"):getValue()
			vco2Rmsb = math.floor(vco2RSliderValue/16)
			vco2Rlsb = vco2RSliderValue%16
			if midiBuffer[9] ~= vco2Rmsb then
				bufferChanged = true
				midiBuffer[9] = vco2Rmsb
			end

-- 500 more lines of code

-- Send out MIDI Messages. Update Buffers. Set Buffer Changed Flag
			if bufferChanged == true then
		
			midiBufferMessage = CtrlrMidiMessage(midiBuffer)
			panel: sendMidiMessageNow(midiBufferMessage)
			bufferChanged = false
			lastMidiBufferMessage = midiBufferMessage
			lastMidiBuffer = midiBuffer
	
			end --if	
			
		
		end
		if panel:getModulatorByName("controlModeButton"):getValue() == 0 then
			timer:stopTimer(timerId)
		else
			return
		end
end
```

On every slider I have a function on value change. 
```
sliderMovedMethod = function(--[[ CtrlrModulator --]] mod, --[[ number --]] value, --[[ number --]] source)
-- This method changes the buffer changed flag, so that the timer loop knows to send a dump when required.
bufferChanged = true

end
```

Hope this helps.

**Reply by damiensellier** [reply]
I totally understand. So this way it holds the modified modulator value in the buffer until the next midiDelayTime in ms?
 I will try something like this and add another trigger for when the DAW reads an automation instead of a mouseDown action.  midiDelayTime is in ms?

**Reply by Godlike-Productions** [reply]
It doesn't really hold it in a buffer.  It reads the state of the modulators at the next midiDelayTime. I guess the modulators are a bit like a buffer. 

I name a variable midiBuffer, but that is only really for assembling the sysex message prior to the send MIDI message. I use a table because it is convenient, but you don't have to. It only needs to exist for the amount of time it takes LUA to sample the values each time the timerCallback is initiated by the timer. It is not updated in the time between the timerCallback calls.  I left it as a global variable though because I use it for an unrelated task when I save the patch to the library, or to a RAM slot on the Jupiter

Yes midiDelayTime is in ms
